### PR TITLE
Porting shape definitions and resampling

### PR DESCRIPTION
**Pull request by wodeni**
# Description

Related issues: #352

This PR ports shape definitions into the frontend (`ShapeDefs.ts`), along with resampling capabilities. This, together with frontend objectives and constraints, will allow energy-based resampling entirely in `penrose-web`.

# Implementation strategy and design decisions

* New `ShapeDefs` module
* Random sampling using a global seed instead of pure generators in HS

# Examples with steps to reproduce them

`penrose set-theory-domain/twosets.sub set-theory-domain/venn-opt-test.sty set-theory-domain/setTheory.dsl`
resample-working
`penrose mesh-set-domain/DomainInterop.sub mesh-set-domain/DomainInterop.sty mesh-set-domain/DomainInterop.dsl`
resample-large

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran [Haddock](https://github.com/penrose/penrose/wiki/Writing-and-generating-documentation#generating-html-documentation-site) and there were no errors when generating the HTML site
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `stack test`
- [x] My code follows the style guidelines of this project (e.g.: no HLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

### CODE REVIEW

**Comment by wodeni** [comment]
All shape definitions are in `shapes/ShapeDef.ts`. Here's one example:
```ts
export const circleDef: ShapeDef = {
  shapeType: "Circle",
  properties: {
    center: ["VectorV", vectorSampler],
    r: ["FloatV", widthSampler],
    strokeWidth: ["FloatV", strokeSampler],
    style: ["StrV", () => constValue("StrV", "filled")],
    strokeStyle: ["StrV", () => constValue("StrV", "solid")],
    strokeColor: ["ColorV", colorSampler],
    color: ["ColorV", colorSampler],
    name: ["StrV", () => constValue("StrV", "defaultCircle")],
  },
};
```

File: penrose-web/src/shapes/ShapeDef.ts [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    throw new Error(
+      `${property} is not a valid property to be sampled for shape ${shapeDef.shapeType}.`
+    );
+};
+
+/**
+ * Sample varying fields, which are assumed to be positional values. They are sampled with the canvas dimensions.
+ * @param state State that contains a list of varying paths
+ */
+const sampleFields = ({ varyingPaths }: State): number[] => {
```
Comment (hypotext) [inline penrose-web/src/shapes/ShapeDef.ts:313]: Should annotate unused functions with `@ignore`

File: penrose-web/src/shapes/ShapeDef.ts [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+  const { varyingPaths, shapes, uninitializedPaths, params } = state;
+  const varyingValues: Value<number>[] = varyingPaths.map((p: Path) =>
+    samplePath(p, shapes)
+  );
+  const uninitValues: Value<VarAD>[] = uninitializedPaths.map((p: Path) =>
+    valueNumberToAutodiff(samplePath(p, shapes))
+  );
+
+  // update the translation with all uninitialized values (converted to `Done` values)
+  const uninitExprs: TagExpr<VarAD>[] = uninitValues.map((v) => val2Expr(v));
```
Comment (hypotext) [inline penrose-web/src/shapes/ShapeDef.ts:351]: Can remove parens on lambda

File: penrose-web/src/shapes/ShapeDef.ts [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+  }
+};
+
+export const resampleOne = (state: State): State => {
+  // resample all the uninitialized and varying values
+  const { varyingPaths, shapes, uninitializedPaths, params } = state;
+  const varyingValues: Value<number>[] = varyingPaths.map((p: Path) =>
+    samplePath(p, shapes)
+  );
+  const uninitValues: Value<VarAD>[] = uninitializedPaths.map((p: Path) =>
```
Comment (hypotext) [inline penrose-web/src/shapes/ShapeDef.ts:346]: Can you remind me why uninitialized paths need to be resampled? I thought it was just the label dimensions, which should be held constant.

**Review by hypotext — APPROVED** [review]
Just some minor comments, looks good overall and can be merged after you address the comments. Thanks!

File: penrose-web/src/shapes/ShapeDef.ts [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+};
+
+//#endregion
+
+//#region shapedefs
+type ShapeDef = IShapeDef;
+
+// type HasTag<T, N> = T extends { tag: N } ? T : never;
+
+type PropType = Value<number>["tag"];
```
[the remark this answers is not in the document]
Comment (wodeni) [reply penrose-web/src/shapes/ShapeDef.ts:91]: https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html

File: penrose-web/src/shapes/ShapeDef.ts (hunk 3 above)
Comment (wodeni) [reply penrose-web/src/shapes/ShapeDef.ts:346]: No, it's a fairly big list of all other properties (e.g. `color`, `strokeColor` etc). You were referring to `pendingProperties`

File: penrose-web/src/shapes/ShapeDef.ts (hunk 2 above)
Comment (wodeni) [reply penrose-web/src/shapes/ShapeDef.ts:351]: Generally, it's encouraged to add type annotations, which involve parens. So I think ESLint is okay with this.

File: penrose-web/src/shapes/ShapeDef.ts [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+/**
+ * A registry of all types of shape definitions in the Penrose system.
+ */
+export const shapedefs: ShapeDef[] = [
+  circleDef,
+  textDef,
+  rectDef,
+  squareDef,
+  curveDef,
```
[the remark this answers is not in the document]
Comment (wodeni) [reply penrose-web/src/shapes/ShapeDef.ts:239]: Is there any working example that uses `Curve`?

File: penrose-web/src/utils/Util.tsx [hunk 6]
```diff
@@ -2,13 +2,24 @@ import * as React from "react";
...
 import memoize from "fast-memoize";
+import { times } from "lodash";
 
 /**
  * Generate a random float. The maximum is exclusive and the minimum is inclusive
  * @param min minimum (inclusive)
  * @param max maximum (exclusive)
  */
-export const randFloat = (min: number, max: number) =>
+export const randFloats = (
```
[the remark this answers is not in the document]
Comment (wodeni) [reply penrose-web/src/utils/Util.tsx:12]: Will do

File: penrose-web/src/shapes/ShapeDef.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (hypotext) [reply penrose-web/src/shapes/ShapeDef.ts:239]: Yes, the `perpMark` in this example:

Orthogonal vector example from paper: `runpenrose linear-algebra-domain/twoVectorsPerp.sub linear-algebra-domain/linear-algebra-paper-simple.sty linear-algebra-domain/linear-algebra.dsl`

**Pull request merged**